src/chainstate/stacks/address.rs

```diff
@@ -21,24 +21,238 @@ use net::StacksMessageCodec;
+        if new_index as usize > buf.len() {
+            return Err(net_error::UnderflowError);
+        }
+
+        let version : u8    = read_next(buf, &mut index, max_size)?;
+        let mut hash_bytes = [0u8; 20];
+        hash_bytes.copy_from_slice(&buf[index as usize..(index+HASH160_ENCODED_SIZE) as usize]);
+       
+        let bytes = Hash160(hash_bytes);
+        index += HASH160_ENCODED_SIZE;
+
+        *index_ptr = index;
+        
         Ok(StacksAddress {
             version: version,
             bytes: bytes
         })
     }
 }
+
+impl StacksAddress {
+    pub fn new(version: u8, hash: Hash160) -> StacksAddress {
+        StacksAddress {
+            version,
+            bytes: hash
+        }
+    }
+
+    /// Generate an address from a given address hash mode, signature threshold, and list of public
+    /// keys.  Only return an address if the combination given is supported.
+    /// The version is may be arbitrary.
+    pub fn from_public_keys(version: u8, hash_mode: &AddressHashMode, num_sigs: usize, pubkeys: &Vec<StacksPublicKey>) -> Option<StacksAddress> {
+        // must be sufficient public keys
+        if pubkeys.len() < num_sigs {
+            return None
+        }
+
+        // address hash mode must be consistent with the number of keys
+        match *hash_mode {
+            AddressHashMode::SerializeP2PKH | AddressHashMode::SerializeP2WPKH => {
+                // must be a single public key, and must require one signature
+                if num_sigs != 1 || pubkeys.len() != 1 {
+                    return None;
+                }
+            },
+            _ => {}
+        }
+
+        // if segwit, then keys must all be compressed
+        match *hash_mode {
+            AddressHashMode::SerializeP2WPKH | AddressHashMode::SerializeP2WSH => {
+                for pubkey in pubkeys {
+                    if !pubkey.compressed() {
+                        return None;
+                    }
+                }
+            },
+            _ => {}
+        }
+        
+        let hash_bits = public_keys_to_address_hash(hash_mode, num_sigs, pubkeys);
+        Some(StacksAddress::new(version, hash_bits))
+    }
+
+    /// Convert from a Bitcoin address
+    pub fn from_bitcoin_address(addr: &BitcoinAddress) -> StacksAddress {
+        let btc_version = address_type_to_version_byte(addr.addrtype, addr.network_id);
+
+        // should not fail by construction
+        let version = to_c32_version_byte(btc_version).expect("Failed to decode Bitcoin version byte to Stacks version byte");
+        StacksAddress {
+            version: version,
+            bytes: addr.bytes.clone()
+        }
+    }
+
+    /// Convert to PrincipalData::Standard(StandardPrincipalData)
+    pub fn to_account_principal(&self) -> PrincipalData {
+        PrincipalData::Standard(StandardPrincipalData(self.version, self.bytes.as_bytes().clone()))
+    }
+}
+
+impl Address for StacksAddress {
+    fn to_bytes(&self) -> Vec<u8> {
```

Comment:
💅 should we stick to rust conventions, and name this method `into_bytes`, instead of `to_bytes`?

Reply:
I think `to_bytes()` means "create a byte representation without destroying the object" whereas `into_bytes()` means "convert the object into its byte representation".  The semantics of `to_bytes()` are desired here.